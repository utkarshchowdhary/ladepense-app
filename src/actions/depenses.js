const { v4: uuidv4 } = require('uuid');

export const addDepense = ({
  description = '',
  note = '',
  amount = 0,
  createdAt = 0,
} = {}) => ({
  type: 'ADD_DEPENSE',
  depense: {
    id: uuidv4(),
    description,
    note,
    amount,
    createdAt,
  },
});

export const removeDepense = ({ id } = {}) => ({
  type: 'REMOVE_DEPENSE',
  id,
});

export const editDepense = (id, updates) => ({
  type: 'EDIT_DEPENSE',
  id,
  updates,
});
