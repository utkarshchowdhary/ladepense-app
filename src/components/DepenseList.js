import React from 'react';
import { connect } from 'react-redux';
import DepenseListItem from './DepenseListItem';
import selectDepenses from '../selectors/depenses';

export const DepenseList = (props) => (
  <div>
    {props.depenses.length === 0 ? (
      <p>No depenses</p>
    ) : (
      props.depenses.map((depense) => (
        <DepenseListItem key={depense.id} {...depense} />
      ))
    )}
  </div>
);

const mapStateToProps = (state) => ({
  depenses: selectDepenses(state.depenses, state.filters),
});

export default connect(mapStateToProps)(DepenseList);
