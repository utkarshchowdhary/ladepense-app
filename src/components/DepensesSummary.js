import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectDepenses from '../selectors/depenses';
import selectDepensesTotal from '../selectors/depenses-total';

export const DepensesSummary = ({ depenseCount, depensesTotal }) => {
  const depenseWord = depenseCount === 1 ? 'depense' : 'depenses';
  const formattedDepensesTotal = numeral(depensesTotal / 100).format('$0,0.00');
  return (
    <div>
      <h1>
        Viewing {depenseCount} {depenseWord} totalling {formattedDepensesTotal}
      </h1>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleDepenses = selectDepenses(state.depenses, state.filters);

  return {
    depenseCount: visibleDepenses.length,
    depensesTotal: selectDepensesTotal(visibleDepenses),
  };
};

export default connect(mapStateToProps)(DepensesSummary);
