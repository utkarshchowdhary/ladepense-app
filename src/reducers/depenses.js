const depensesReducerDefaultState = [];

export default (state = depensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_DEPENSE':
      return [...state, action.depense];
    case 'REMOVE_DEPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_DEPENSE':
      return state.map((depense) => {
        if (depense.id === action.id) {
          return {
            ...depense,
            ...action.updates,
          };
        } else {
          return depense;
        }
      });
    default:
      return state;
  }
};
