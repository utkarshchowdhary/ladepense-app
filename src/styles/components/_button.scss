.button {
  background: $blue;
  border: none;
  color: #fff;
  display: inline-block;
  font-size: $font-size-large;
  font-weight: 300;
  line-height: 1;
  padding: $s-size;
  text-decoration: none;
}

.button--link {
  background: none;
  border: 0.1rem solid #fff;

  &:hover {
    background: #fff;
    color: $midnight-blue;
  }
}

.button--secondary {
  background: $slate-grey;
}
