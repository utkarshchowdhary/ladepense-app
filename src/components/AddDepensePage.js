import React from 'react';
import { connect } from 'react-redux';
import { addDepense } from '../actions/depenses';
import DepenseForm from './DepenseForm';

export class AddDepensePage extends React.Component {
  onSubmit = (depense) => {
    this.props.addDepense(depense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Add Depense</h1>
        <DepenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  addDepense: (depense) => dispatch(addDepense(depense)),
});

export default connect(undefined, mapDispatchToProps)(AddDepensePage);
