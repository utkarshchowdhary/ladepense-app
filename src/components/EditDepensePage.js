import React from 'react';
import { connect } from 'react-redux';
import { editDepense, removeDepense } from '../actions/depenses';
import DepenseForm from './DepenseForm';

export class EditDepensePage extends React.Component {
  onSubmit = (depense) => {
    this.props.editDepense(this.props.depense.id, depense);
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.removeDepense(this.props.depense.id);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <DepenseForm depense={this.props.depense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  depense: state.depenses.find(
    (depense) => depense.id === props.match.params.id
  ),
});

const mapDispatchToProps = (dispatch) => ({
  editDepense: (id, depense) => dispatch(editDepense(id, depense)),
  removeDepense: (id) => dispatch(removeDepense({ id })),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditDepensePage);
