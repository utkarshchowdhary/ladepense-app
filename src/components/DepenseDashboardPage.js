import React from 'react';
import DepenseList from './DepenseList';
import DepenseListFilters from './DepenseListFilters';

const DepenseDashboardPage = () => (
  <div>
    <DepenseListFilters />
    <DepenseList />
  </div>
);

export default DepenseDashboardPage;
