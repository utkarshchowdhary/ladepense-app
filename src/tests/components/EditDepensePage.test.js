import React from 'react';
import { shallow } from 'enzyme';
import { EditDepensePage } from '../../components/EditDepensePage';
import depenses from '../fixtures/depenses';

let editDepense, removeDepense, history, wrapper;

beforeEach(() => {
  editDepense = jest.fn();
  removeDepense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditDepensePage
      editDepense={editDepense}
      removeDepense={removeDepense}
      history={history}
      depense={depenses[2]}
    />
  );
});

test('should render EditDepensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle editDepense', () => {
  wrapper.find('DepenseForm').prop('onSubmit')(depenses[2]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editDepense).toHaveBeenLastCalledWith(depenses[2].id, depenses[2]);
});

test('should handle removeDepense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeDepense).toHaveBeenLastCalledWith(depenses[2].id);
});
