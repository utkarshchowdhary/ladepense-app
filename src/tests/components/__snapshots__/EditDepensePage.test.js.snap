// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditDepensePage correctly 1`] = `
<div>
  <DepenseForm
    depense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
